python/feathub/processors/tests/feathub_test_base.py

```diff
@@ -0,0 +1,157 @@
+    of test cases.
+    """
+
+    # By setting this attribute to false, it prevents pytest from discovering
+    # this class as a test when searching up from its child classes.
+    __test__ = False
+
+    def setUp(self) -> None:
+        self.temp_dir = tempfile.mkdtemp()
+        self.input_data, self.schema = self._create_input_data_and_schema()
+        self.client = self.get_client()
+
+    def tearDown(self) -> None:
+        MemoryOnlineStore.get_instance().reset()
+        shutil.rmtree(self.temp_dir, ignore_errors=True)
+
+    @abstractmethod
+    def get_client(self) -> FeathubClient:
+        """
+        Returns a FeathubClient instance for test cases.
+        """
+        pass
+
+    @staticmethod
+    def _get_local_client(processor_config: Dict) -> FeathubClient:
+        return FeathubClient(
+            props={
+                "processor": processor_config,
+                "online_store": {
+                    "types": ["memory"],
+                    "memory": {},
+                },
+                "registry": {
+                    "type": "local",
+                    "local": {
+                        "namespace": "default",
+                    },
+                },
+                "feature_service": {
+                    "type": "local",
+                    "local": {},
+                },
+            }
+        )
+
+    def _create_file_source(
+        self,
+        df: pd.DataFrame,
+        keys: Optional[List[str]] = None,
+        schema: Optional[Schema] = None,
+        timestamp_field: Optional[str] = "time",
+        timestamp_format: str = "%Y-%m-%d %H:%M:%S",
+        name: str = None,
+    ) -> FileSystemSource:
+        path = tempfile.NamedTemporaryFile(dir=self.temp_dir).name
+        if schema is None:
+            schema = self._create_input_schema()
+        df.to_csv(path, index=False, header=False)
+
+        if name is None:
+            name = self._generate_random_name("source")
+
+        return FileSystemSource(
+            name=name,
+            path=path,
+            data_format="csv",
+            schema=schema,
+            keys=keys,
+            timestamp_field=timestamp_field,
+            timestamp_format=timestamp_format,
+        )
+
+    def _create_input_data_and_schema(self):
+        input_data = pd.DataFrame(
+            [
+                ["Alex", 100, 100, "2022-01-01 08:01:00"],
+                ["Emma", 400, 250, "2022-01-01 08:02:00"],
+                ["Alex", 300, 200, "2022-01-02 08:03:00"],
+                ["Emma", 200, 250, "2022-01-02 08:04:00"],
+                ["Jack", 500, 500, "2022-01-03 08:05:00"],
+                ["Alex", 600, 800, "2022-01-03 08:06:00"],
+            ],
+            columns=["name", "cost", "distance", "time"],
+        )
+
+        schema = self._create_input_schema()
+
+        return input_data, schema
+
+    def _create_input_data_and_schema_with_millis_time_span(self):
```

Comment:
It seems inconsistent that `_create_input_schema` is marked as static but this method is not.